src/borrow_tracker/tree_borrows/wildcard.rs

```diff
@@ -0,0 +1,409 @@
+/// Note that we derive Ord and PartialOrd, so the order in which variants are listed below matters:
+/// None < Read < Write. Do not change that order. See the `test_order` test.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Default)]
+pub enum WildcardAccessLevel {
+    #[default]
+    None,
+    Read,
+    Write,
+}
+/// were relative to the pointer the access happened from
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum WildcardAccessRelatedness {
+    /// the access definitively happened through a child pointer
+    ChildAccess,
+    /// the access definitively happened through a foreign pointer
+    ForeignAccess,
+    /// we do not know if the access is foreign or child
+    EitherAccess,
+}
+impl WildcardAccessRelatedness {
+    pub fn to_relatedness(self) -> Option<AccessRelatedness> {
+        match self {
+            Self::ChildAccess => Some(AccessRelatedness::WildcardChildAccess),
+            Self::ForeignAccess => Some(AccessRelatedness::WildcardForeignAccess),
+            Self::EitherAccess => None,
+        }
+    }
+}
+
+/// state per location per pointer keeping track of where relative to this
+/// node exposed pointers are and what access permissions they have
+///
+/// designed to be completely determined by its parent, siblings and direct
+/// childrens max_child_access/max_foreign_access
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct WildcardAccessTracking {
+    /// how many of this nodes direct children have `max_child_access==Write`
+    child_writes: u16,
+    /// how many of this nodes direct children have `max_child_access>=Read`
+    child_reads: u16,
+    /// the maximum access level that could happen from an exposed
+    /// pointer that is foreign to this pointer
+    ///
+    /// this is calculated as the `max()` of parents `max_foreign_access`, `exposed_as` and
+    /// its siblings `max_child_access`
+    max_foreign_access: WildcardAccessLevel,
+}
+impl WildcardAccessTracking {
+    /// the maximum access level that could happen from an exposed
+    /// pointer that is a child of this pointer
+    pub fn max_child_access(&self, exposed_as: WildcardAccessLevel) -> WildcardAccessLevel {
+        use WildcardAccessLevel::*;
+        max(
+            exposed_as,
+            if self.child_writes > 0 {
+                Write
+            } else if self.child_reads > 0 {
+                Read
+            } else {
+                None
+            },
+        )
+    }
+    /// from where relative to this pointer a read or write access could happen
+    pub fn access_relatedness(
+        &self,
+        kind: AccessKind,
+        exposed_as: WildcardAccessLevel,
+    ) -> Option<WildcardAccessRelatedness> {
+        match kind {
+            AccessKind::Read => self.read_access_relatedness(exposed_as),
+            AccessKind::Write => self.write_access_relatedness(exposed_as),
+        }
+    }
+    /// from where relative to this pointer a read access could happen
```

Comment:
"this pointer"/"this node" is confusing here since you don't actually identify a node here.
What you mean is "a node with this wildcard info" or so?